wandb/sdk/wandb_init.py

```diff
@@ -138,20 +124,18 @@ def setup(self, kwargs):
         if magic not in (None, False):
             magic_install(kwargs)
 
-        # prevent setting project, entity if in sweep
-        # TODO(jhr): these should be locked elements in the future or at least
-        #            moved to apply_init()
-        if settings.sweep_id:
-            for key in ("project", "entity"):
-                val = kwargs.pop(key, None)
-                if val:
-                    print("Ignored wandb.init() arg %s when running a sweep" % key)
-        settings.apply_init(kwargs)
-
-        login_key = wandb_login._login(_disable_warning=True, _settings=settings)
+        # handle login related parameters as these are applied to global state
+        anonymous = kwargs.pop("anonymous", None)
+        force = kwargs.pop("force", None)
+        login_key = wandb.login(anonymous=anonymous, force=force)
         if not login_key:
             settings.mode = "offline"
 
+        # apply updated global state after login was handled
+        settings._apply_settings(wandb.setup()._settings)
+
+        settings._apply_init(kwargs)
+
```

Comment:
just making sure I understand: the login stuff modifies the global settings, which you've previously cloned, so now you need to pull the changes to the globals back in?

backing up a bit more, is the point of putting the global settings object in the wandbsetup object to have some base settings object that can be extended/modified for each run in a session?

Reply:
yes.  but only bringing the right things back into the singleton object,  Right now it is api_key, force, and anonymous.   It is still a bit error prone how it happens (too manual) im trying to come up with a better abstraction that manages this logic (probably in settings)
So we will just annotate certain settings as global, and they should automatically be applied to the singleton object.